File: Wire-iOS/Sources/UserInterface/Components/Views/BackgroundView/BackgroundViewController.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "BackgroundViewController.h"
#import "UserBackgroundView.h"
#import "zmessaging+iOS.h"
@import WireExtensionComponents;

#import "UIColor+WAZExtensions.h"
#import "UIView+Borders.h"
#import "Wire-Swift.h"
#import "WAZUIMagic.h"
#import "AccentColorChangeHandler.h"
#import "Constants.h"
#import "UIView+WR_ExtendedBlockAnimations.h"
#import "ColorSchemeController.h"

@interface BackgroundViewController (ConversationObserver) <ZMConversationObserver>

- (void)updateBackgroundForConversation;

@end

@interface BackgroundViewController ()

@property (nonatomic, strong) id<ZMBareUser> user;
@property (nonatomic, strong) id<ZMBareUser> overrideUser;

@property (nonatomic, strong) UIColor *filterColor;

@property (nonatomic, strong) UserBackgroundView *backgroundView;
@property (nonatomic, strong) NSLayoutConstraint *backgroundViewFullScreenConstraint;
@property (nonatomic, strong) NSLayoutConstraint *backgroundViewSidebarConstraint;

@property (nonatomic, strong) AccentColorChangeHandler *accentColorHandler;
@property (nonatomic) id conversationObserverToken;

@end



@implementation BackgroundViewController

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    UIColor *accentColor = [ZMUser selfUser].accentColor;
    if (! accentColor) {
        DDLogWarn(@"User has no accent color, picking ZMAccentColorSoftPink");
        accentColor = [UIColor colorForZMAccentColor:ZMAccentColorSoftPink];
    }
    
    self.backgroundView = [[UserBackgroundView alloc] initWithFilterColor:accentColor];
    self.backgroundView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.backgroundView];

    @weakify(self);
    self.accentColorHandler = [AccentColorChangeHandler addObserver:self handlerBlock:^(UIColor *newColor, id object) {
        @strongify(self);
        self.filterColor = newColor;
        
        if (! self.overrideFilterColor) {
            self.backgroundView.filterColor = newColor;
        }
    }];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(colorSchemeControllerDidApplyChanges:) name:ColorSchemeControllerDidApplyColorSchemeChangeNotification object:nil];
    
    [self.backgroundView addConstraintsVerticallyFittingToView:self.view];
    [self.backgroundView addConstraintForLeftMargin:0.0 relativeToView:self.view];
    
    self.backgroundViewFullScreenConstraint = [self.backgroundView addConstraintForRightMargin:0.0 relativeToView:self.view];

    CGFloat sidebarWidth = IS_IPAD ? [WAZUIMagic cgFloatForIdentifier:@"framework.sidebar_width"] : [UIScreen mainScreen].bounds.size.width;
    self.backgroundViewSidebarConstraint = [self.backgroundView addConstraintForWidth:sidebarWidth];
    [self updateBackgroundViewLayout];
}

- (void)updateBackgroundViewLayout
{
    if (self.traitCollection.userInterfaceIdiom == UIUserInterfaceIdiomPad &&  ! self.forceFullScreen) {
        self.backgroundViewFullScreenConstraint.active = NO;
        self.backgroundViewSidebarConstraint.active = YES;
    }
    else {
        self.backgroundViewSidebarConstraint.active = NO;
        self.backgroundViewFullScreenConstraint.active = YES;
    }
}

- (void)setForceFullScreen:(BOOL)forceFullScreen
{
    [self setForceFullScreen:forceFullScreen animated:NO];
}

- (void)setForceFullScreen:(BOOL)forceFullScreen animated:(BOOL)animated
{
    _forceFullScreen = forceFullScreen;

    void (^animationBlock)() = ^() {
        [self updateBackgroundViewLayout];
    };
    
    if (animated) {
        [UIView wr_animateWithEasing:RBBEasingFunctionEaseOutExpo 
                            duration:[WAZUIMagic floatForIdentifier:@"background.animation_duration"] 
                          animations:^{
                              animationBlock(); 
                              [self updateViewConstraints];
                              [self.view layoutIfNeeded];
                          }];        
    }
    else {
        animationBlock();
    }
}

- (CGFloat)blurPercent
{
    return self.backgroundView.blurPercent;
}

- (void)setBlurPercent:(CGFloat)blurPercent
{
    self.backgroundView.blurPercent = blurPercent;
}

- (void)setBlurPercentAnimated:(CGFloat)blurPercent
{
    [self.backgroundView setBlurPercentAnimated:blurPercent];
}

- (BOOL)blurDisabled
{
    return self.backgroundView.blurDisabled;
}

- (void)setBlurDisabled:(BOOL)blurDisabled
{
    self.backgroundView.blurDisabled = blurDisabled;
}

- (void)setUser:(id<ZMBareUser>)user animated:(BOOL)animated
{
    _user = user;
    
    if (! self.overrideUser) {
        [self.backgroundView setUser:user animated:animated waitForBlur:YES];
    }
}

- (void)setOverrideUser:(id<ZMBareUser>)overrideUser disableColorFilter:(BOOL)disableColorFilter animated:(BOOL)animated;
{
    [self setOverrideUser:overrideUser disableColorFilter:(BOOL)disableColorFilter animated:animated completionBlock:nil];
}

- (void)setOverrideUser:(id<ZMBareUser>)overrideUser disableColorFilter:(BOOL)disableColorFilter animated:(BOOL)animated completionBlock:(dispatch_block_t)completionBlock;
{
    _overrideUser = overrideUser;
    
    if (overrideUser == nil) {
        [self unsetOverrideUserAnimated:animated];
    }
    else {
        [self.backgroundView setUser:overrideUser animated:animated waitForBlur:YES];
        self.backgroundView.filterDisabled = disableColorFilter;
    }
    
    if (completionBlock != nil) {
        completionBlock();
    }
}

- (void)unsetOverrideUserAnimated:(BOOL)animated
{
    self.overrideUser = nil;
    self.backgroundView.filterDisabled = NO;
    [self.backgroundView setUser:self.user animated:animated waitForBlur:YES];
}

- (void)setOverrideFilterColor:(UIColor *)overrideFilterColor;
{
    _overrideFilterColor = overrideFilterColor;
    
    if (overrideFilterColor != nil) {
        self.backgroundView.filterColor = overrideFilterColor;
    } else {
        self.backgroundView.filterColor = self.filterColor;
    }
}

- (void)setConversation:(ZMConversation *)conversation
{
    if (_conversation != conversation) {
        _conversation = conversation;
        if (_conversation != nil) {
            self.conversationObserverToken = [ConversationChangeInfo addObserver:self forConversation:self.conversation];
        }
        [self updateBackgroundForConversation];
    }
}

#pragma mark - ColorSchemeControllerDidApplyChangesNotification

- (void)colorSchemeControllerDidApplyChanges:(NSNotification *)notification
{
    [self.backgroundView updateAppearanceAnimated:YES];
}

@end

@implementation BackgroundViewController (ConversationObserver)

- (void)conversationDidChange:(ConversationChangeInfo *)change
{
    if (change.messagesChanged) {
        [self updateBackgroundForConversation];
    }
}

- (void)updateBackgroundForConversation
{
    ZMUser *user = self.conversation.lastMessageSender;
    if (user) {
        [self setUser:user animated:YES];
    }
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/BackgroundViewController.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import UIKit
import WireSyncEngine

final class BackgroundViewController: UIViewController {

    // MARK: - Constants

    private let backgroundScaleFactor: CGFloat = 1.4

    // MARK: - Properties

    var accentColor: UIColor {
        get { imageView.backgroundColor ?? .clear }
        set { imageView.backgroundColor = newValue }
    }

    // MARK: - Private Properties

    private let imageView = UIImageView()
    private let cropView = UIView()
    private let darkenOverlay = UIView()
    private let blurView = UIVisualEffectView(effect: UIBlurEffect(style: .dark))
    private let imageTransformer: ImageTransformer

    // MARK: - Life Cycle

    init(
        accentColor: UIColor,
        imageTransformer: ImageTransformer
    ) {
        self.imageTransformer = imageTransformer
        super.init(nibName: .none, bundle: .none)
        self.accentColor = accentColor
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        configureViews()
        createConstraints()
    }

    private func configureViews() {

        imageView.contentMode = .scaleAspectFill
        imageView.transform = CGAffineTransform(scaleX: backgroundScaleFactor, y: backgroundScaleFactor)

        cropView.clipsToBounds = true
        [imageView, blurView, darkenOverlay].forEach(cropView.addSubview)
        view.addSubview(cropView)
    }

    private func createConstraints() {
        cropView.translatesAutoresizingMaskIntoConstraints = false
        blurView.translatesAutoresizingMaskIntoConstraints = false
        imageView.translatesAutoresizingMaskIntoConstraints = false
        darkenOverlay.translatesAutoresizingMaskIntoConstraints = false

        let constraints = [
            // Crop view
            cropView.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: -100),
            cropView.topAnchor.constraint(equalTo: view.topAnchor),
            cropView.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: 100),
            cropView.bottomAnchor.constraint(equalTo: view.bottomAnchor),

            // Blur view
            blurView.leadingAnchor.constraint(equalTo: cropView.leadingAnchor),
            blurView.topAnchor.constraint(equalTo: cropView.topAnchor),
            blurView.trailingAnchor.constraint(equalTo: cropView.trailingAnchor),
            blurView.bottomAnchor.constraint(equalTo: cropView.bottomAnchor),

            // Image view
            imageView.leadingAnchor.constraint(equalTo: cropView.leadingAnchor),
            imageView.topAnchor.constraint(equalTo: cropView.topAnchor),
            imageView.trailingAnchor.constraint(equalTo: cropView.trailingAnchor),
            imageView.bottomAnchor.constraint(equalTo: cropView.bottomAnchor),

            // Darken overlay
            darkenOverlay.leadingAnchor.constraint(equalTo: cropView.leadingAnchor),
            darkenOverlay.topAnchor.constraint(equalTo: cropView.topAnchor),
            darkenOverlay.trailingAnchor.constraint(equalTo: cropView.trailingAnchor),
            darkenOverlay.bottomAnchor.constraint(equalTo: cropView.bottomAnchor)
        ]

        NSLayoutConstraint.activate(constraints)
    }

    override var childForStatusBarStyle: UIViewController? { children.first }
    override var childForStatusBarHidden: UIViewController? { children.first }

    // MARK: - Methods

    func setBackgroundImage(_ image: UIImage?) async {
        guard let image else { return imageView.image = nil }

        imageView.image = await Task.detached(priority: .background) { [imageTransformer] in
            imageTransformer.adjustInputSaturation(value: 2, image: image)
        }.value
    }
}
